beacon-chain/utils/shuffle_test.go

```diff
@@ -74,19 +66,105 @@ func TestSplitIndices_OK(t *testing.T) {
 	}
 }
 
-func BenchmarkPermuteIndex(b *testing.B) {
+func TestShuffleList_Vs_ShuffleIndex(t *testing.T) {
+	list := []uint64{}
+	listSize := uint64(1000)
+	seed := [32]byte{123, 42}
+	for i := uint64(0); i < listSize; i++ {
+		list = append(list, i)
+	}
+	shuffledListByIndex := make([]uint64, listSize)
+	for i := uint64(0); i < listSize; i++ {
+		si, err := ShuffledIndex(i, listSize, seed)
+		if err != nil {
+			t.Error(err)
+		}
+		shuffledListByIndex[si] = i
+	}
+	shuffledList, err := ShuffleList(list, seed)
+	if err != nil {
+		t.Fatalf("shuffled list error: %v", err)
+
+	}
+	if !reflect.DeepEqual(shuffledList, shuffledListByIndex) {
+		t.Errorf("shuffled lists ar not equal shuffled list: %v shuffled list by index: %v", shuffledList, shuffledListByIndex)
+	}
+
+}
+
+func BenchmarkShuffledIndex(b *testing.B) {
 	listSizes := []uint64{4000000, 40000, 400}
-	// Random 32 bytes seed for testing.
 	seed := [32]byte{123, 42}
 	for _, listSize := range listSizes {
-		b.Run(fmt.Sprintf("PermuteIndex_%d", listSize), func(ib *testing.B) {
+		b.Run(fmt.Sprintf("ShuffledIndex_%d", listSize), func(ib *testing.B) {
 			for i := uint64(0); i < uint64(ib.N); i++ {
-				PermutedIndex(i%listSize, listSize, seed)
+				ShuffledIndex(i%listSize, listSize, seed)
+			}
+		})
+	}
+}
+
+func BenchmarkIndexComparison(b *testing.B) {
+	listSizes := []uint64{400000, 40000, 400}
+	seed := [32]byte{123, 42}
+	for _, listSize := range listSizes {
+		b.Run(fmt.Sprintf("Indexwise_ShuffleList_%d", listSize), func(ib *testing.B) {
+			for i := 0; i < ib.N; i++ {
+				// Simulate a list-shuffle by running shuffle-index listSize times.
+				for j := uint64(0); j < listSize; j++ {
+					ShuffledIndex(j, listSize, seed)
+				}
+			}
+		})
+	}
+}
+
+func BenchmarkShuffleList(b *testing.B) {
+	listSizes := []uint64{400000, 40000, 400}
+	seed := [32]byte{123, 42}
+	for _, listSize := range listSizes {
+		testIndices := make([]uint64, listSize, listSize)
+		for i := uint64(0); i < listSize; i++ {
+			testIndices[i] = i
+		}
+		b.Run(fmt.Sprintf("ShuffleList_%d", listSize), func(ib *testing.B) {
+			for i := 0; i < ib.N; i++ {
+				ShuffleList(testIndices, seed)
 			}
 		})
 	}
 }
 
+func TestShuffledIndex(t *testing.T) {
+	list := []uint64{}
+	listSize := uint64(400)
+	for i := uint64(0); i < listSize; i++ {
+		list = append(list, i)
+	}
+	shuffledList := make([]uint64, listSize)
+	unShuffledList := make([]uint64, listSize)
+	seed := [32]byte{123, 42}
+	for i := uint64(0); i < listSize; i++ {
+		si, err := ShuffledIndex(i, listSize, seed)
+		if err != nil {
+			t.Error(err)
+		}
+		shuffledList[si] = i
+	}
+	t.Logf("shuffledList: %v", shuffledList)
+	for i := uint64(0); i < listSize; i++ {
+		ui, err := UnShuffledIndex(i, listSize, seed)
```

Comment:
This test doesnt look right, why are you shuffling and unshuffling on the same index ? shouldnt you unshuffle the shuffled list ?

Reply:
not readable but it was doing the same. changed it

Reply:
checking again

Reply:
i think it is ok it re arrange the shuffled array by the un-shuffle index